Allow BITS to retry downloads

wjt opened the pull request:
The changes are, roughly:

* Don't terminate downloads when the connection state changes
* Treat TRANSIENT_ERROR for image downloads as "the download is proceeding", not as if it were ERROR
* Show an error on TRANSIENT_ERROR for the JSON download, but keep retrying
* Don't set absurdly low timeouts on BITS jobs
* Ensure we always notify the UI when a real error occurs (a bug masked by all the above)
* drop the JSONDownloadFailed error state entirely, since it is no longer a "jump-to-the-error-page" kind of error, and add a separate analytics event for that case to track its frequency

https://phabricator.endlessm.com/T14535

ramcq committed:
refactor JobModification to avoid many levels of nesting

ramcq committed:
don't handle TRANSIENT_ERROR as fatal

ramcq committed:
don't set NoProgressTimeout on BITS jobs

Leave the current, very large, default value of 14 days before BITS will give
up on downloading a file and proceed from TransientError to Error.

(Incidentally, the #define values were not used, meaning that any transient
error would be upgraded immediately to a permanent error by BITS.)

ramcq committed:
move to online without any additional checking beyond InternetGetConnectedState

Now that BITS retries transient errors for us, don't do any additional
checking.

wjt committed:
DownloadManager: consistent "Job $name $STATE" messages

wjt committed:
Refactor updating "can download" state

wjt committed:
StartJSONDownload: reuse existing job if present

This prevents a spurious failure if the job already exists. Previously,
that job would be cancelled, which would signal an error from the job,
which would cause the app to consider the download to have failed.

wjt committed:
Add FUNCTION_ENTER_FMT macro

I often wish that the FUNCTION_ENTER entries in the log included the
arguments, so here's a macro to make that a bit easier.

wjt committed:
Make JSON download failures non-fatal

Either we have got past the image selection screen (in which case we don't
care about the JSON), or we have not (in which case that screen has a nice
"internet problems" message we can show, rather than the final, scary
error screen with a misleading message).

Ideally we would retry but I think this is good enough: we allow BITS to
retry for a long time, and we retry whenever the internet connection
flaps.

wjt committed:
Do not cancel downloads when the connection drops

BITS' documentation is ambiguous on this point but in practice, if you
unplug the network cable, the transfer goes into state TRANSIENT_ERROR and
the speed gradually decreases; and when you plug it back in again, it
resumes of its own accord.

https://phabricator.endlessm.com/T14535

wjt committed:
Simplify DownloadManager::JobError

The block which switches the job to BG_JOB_PRIORITY_FOREGROUND was
redundant:

* Our files are basically always >2GiB
* We always use BG_JOB_PRIORITY_FOREGROUND anyway, for speedier downloads

CComHeapPtr is neater than manually calling CoTaskMemFree.

In practice GetErrorDescription always fails because the message catalogue
in question is not loaded. I do not fully understand how that is meant to
work, but this means that in practice pszErrorDescription is always NULL,
so we need to deal with that.

wjt committed:
DownloadManager: report errors from JobError

I've seen cases, apparently random, where JobModification is not called
for BG_JOB_STATE_ERROR. The docs do say that JobError is called from a
higher-priority thread so events may appear out of sequence. I think
what happens is the following race:

    JobModification called
    GetState()
    -> TRANSIENT_ERROR
    [ BITS decides the error is not transient, but fatal: ]
    [ preempted ]             JobError called
                              [ ... does its thing ... ]
                              JobError returns
    [ scheduled ]
    JobModification returns
    [ No state change since the last time JobModification returned, so ]
    [ it is not called again!                                          ]

JobError is guaranteed to be called on error, so let's report errors
from there (rather than it being a no-op) and ignore them in
JobModification. In the process, I simplified JobModification a little
bit.

wjt committed:
GetDownloadProgress: ignore TRANSIENT_ERROR

We need to be consistent about this. Otherwise, the thread which polls for
the download status will terminate prematurely.

wjt committed:
Don't show "16384 Pb" when download is starting.

The astute will realise that this is 2 ** 64 bytes.

wjt committed:
Respawn UpdateDownloadProgressThread after an error

This thread is only started if m_downloadUpdateThread ==
INVALID_HANDLE_VALUE, but nothing ever set it back to that value if the
thread terminated.

I'm not sure if this is quite right -- are we leaking the handle to the
old thread? -- but it is strictly better than the current behaviour!

wjt committed:
Send a "fake" ErrorCauseJSONDownloadFailed; remove it from enum

As of the rest of the refactoring, this is not really an "error cause" any
more -- ErrorCause_t describes reasons why the user hits the ErrorPage.

But to avoid cluttering the analytics with a different representation of
roughly the same event, let's represent this (transient, non-fatal) state
the same way.

ramcq reviewed (commented):
I don't quite understand the need for the SetNoProgressTimeout change in 9a8a32a. If we allow this to time out, we are no longer correctly relying on the BITS retry functionality to act as our yardstick of "downloading the JSON successfully means the internet works for us" which was the reason for setting out down this path.

Provided that the user has not yet hit "Install Endless" they will never see whether or not it has timed out or not, so we should either retry ~infinitely in a loop or let BITS do it for us. If they are on the "select image" page but have not succeeded in getting the JSON yet, they will see the "you are not connected" error in perpetuity if the error is caused by something else other than a real download failure (proxy, etc). Either they see the error and bounce their connection and we come to life, or they see it and decide we just suck at the internet and ragequit, but if they have _not_ seen it, I'd prefer we keep patiently retrying so that if the user would, eg, return to their PC then log in to the wifi portal 30 mins later, they would still get the JSON better-late-than-never and be able to proceed.

Unless I am missing something, the only thing we lose by not failing earlier here is the clear metric that we didn't, after some time, manage to download the JSON - but this is already (as you patched elsewhere) a synthetic error. I think we could collect this data less invasively - for example, we can report an event like "showed the image choices to the user" which would be fired after we have  move to the image selection screen and the JSON has been downloaded and parsed. Those events would either appear in succession, or drop-offs between "image screen" and "showing images" would imply JSON failures.

wjt commented:
At present, the “you're not connected” error is only shown when the index download has actually failed. We can change this so that the error is always shown until the index is fetched; or add a synthetic timeout n seconds after the app launches which makes the error appear if the index is not yet fetched.

ramcq commented:
Hmm... can we add a new state, "Retrying" which we enter on the first transient error, then show the error if we have Failed or gone into Retrying, but not actually stop trying in the latter case?

wjt committed:
Introduce a "Retrying" state for JSON download

Per https://github.com/endlessm/rufus/pull/59#pullrequestreview-15188058
we want to show the "something's wrong with the connection" message when
there's a transient error fetching the JSON download, but to keep retrying
essentially forever.

In the process of introducing DownloadStatus_t.transientError, arrange for
UpdateDownloadProgressThread/GetDownloadProgress to always fully
initialize the struct.

wjt committed:
Send JSONDownloadSucceeded event

We'd expect this number to roughly equal the number of new sessions.

wjt commented:
Alright, I removed the offending commit, and arranged for the error message to be displayed if we see any transient errors. If I drop all packets destined for CloudFront it still takes BITS several minutes to report a transient error so I don't think we'll see many false positives (famous last words).

I was going to try to arrange to have a "showed the images to the user" event but it's tedious to arrange for it to be dispatched when the screen is displayed, or when the manifest arrives, whichever comes later; so I chucked in a "success" event as better than nothing
